### PR TITLE
feat(Notion): Added Notion provider.

### PR DESCRIPTION
**Pull request by synchronizing**
# Submitting Pull Requests

## General

 - [x] Make sure you use [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
       Examples: `"fix(google): Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
 - [x] All Python code must formatted using Black, and clean from pep8 and isort issues.
 - [x] JavaScript code should adhere to [StandardJS](https://standardjs.com).
 - [x] If your changes are significant, please update `ChangeLog.rst`.
 - [x] If your change is substantial, feel free to add yourself to `AUTHORS`.

 ## Provider Specifics

 In case you add a new provider:

- [x] Make sure unit tests are available.
- [x] Add an entry of your provider in `test_settings.py::INSTALLED_APPS` and `docs/installation.rst::INSTALLED_APPS`.
- [x] Add documentation to `docs/providers.rst`.
- [x] Add an entry to the list of supported providers over at `docs/overview.rst`.

### CODE REVIEW

**Comment by synchronizing** [comment]
Closes #3112.

**Comment by synchronizing** [comment]
@pennersr Any interest on this addition?

File: allauth/socialaccount/providers/notion/tests.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            reverse(self.provider.id + "_login")
+            + "?"
+            + urlencode(dict(process=process))
+        )
+
+        p = urlparse(resp["location"])
+        q = parse_qs(p.query)
+
+        # Always false.
+        pkce_enabled = False
```
Comment (pennersr) [inline allauth/socialaccount/providers/notion/tests.py:75]: **Question**: There are a few `if pkce_enabled` down below, but here it is hardcoded to false.  Is that intentional?

**Comment by pennersr** [comment]
> @pennersr Any interest on this addition?

Definitely!

File: allauth/socialaccount/providers/notion/tests.py (hunk 1 above)
Comment (synchronizing) [reply allauth/socialaccount/providers/notion/tests.py]: Notion doesn't support PKCE. Looked around when I first implemented this, but did not find a decent way to properly disabled its test in the old codebase, so brought over the `login` test and manually set `pkce_enabled` to `False` with minor mods to get it working.

I see that the `OAuthTestsMixin.login` has been modified in the last few weeks, and just tried 

```python
class NotionTests(OAuth2TestsMixin, TestCase):
    provider_id = NotionProvider.id
    pkce_enabled_default = False
```

with no luck either due to `OAuth2TestMixin.test_provider_has_pkce_params` which overwrites `self.provider.pkce_enabled_default`. Unsure if this is a bug or intended. 

Whatever the case, just committed a new change that cleans up the function a bit more and removes `pkce_enabled` entirely, while also adding `pkce_enabled_default = False` as a class variable. If you think there are any better way to go about this do let me know!

File: allauth/socialaccount/providers/notion/client.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        return "%s?%s" % (authorization_url, urlencode(params))
+
+    def get_access_token(self, code, pkce_code_verifier=None):
+        resp = requests.request(
+            self.access_token_method,
+            self.access_token_url,
+            auth=HTTPBasicAuth(self.consumer_key, self.consumer_secret),
+            json={"code": code, "grant_type": "authorization_code"},
+            headers=self.headers,
+        )
```
Comment (pennersr) [inline allauth/socialaccount/providers/notion/client.py:32]: Can you add a `resp.raise_for_status()` here?

**Review by pennersr — CHANGES_REQUESTED** [review]

File: allauth/socialaccount/providers/notion/client.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        access_token = None
+        if resp.status_code in [200, 201]:
```
[the remark this answers is not in the document]
Comment (synchronizing) [reply allauth/socialaccount/providers/notion/client.py:34]: Is this being caught downstream? I used other `clients.py` as a reference for this block of code (`AppleOAuth2Client.get_access_token` and `OAuthClient.get_access_token`), assuming that `OAuth2Error` is being caught earlier in the call stack for some purposes - unless it's a general `try/except Exception` case. A quick search reveals there is no `raise_for_status` in a `client.py` implementation in the current codebase.

[the remark this answers is not in the document]
Comment (synchronizing) [reply allauth/socialaccount/providers/notion/client.py:34]: Friendly ping @pennersr. 😃

[the remark this answers is not in the document]
Comment (pennersr) [reply allauth/socialaccount/providers/notion/client.py:34]: You're right, my bad.

**Comment by pennersr** [comment]
There are conflicts -- can you please rebase?

**Comment by pennersr** [comment]
> Merged in main.

There still seem to be conflicts...

**Comment by synchronizing** [comment]
> > Merged in main.
> 
> There still seem to be conflicts...

Don't see it, tbh. In my branch seeing 

> This branch is 11 commits ahead of pennersr:main.

Branch looks to be in sync with your main. Coveralls went down 0.2% which caused the check failure - tests look all good. Everything else (other than your change request) looks ok. If you don't mind directing me towards where you see conflicts, I would appreciate it. 😄

**Pull request merged**